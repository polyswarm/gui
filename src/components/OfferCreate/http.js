import web3Utils from 'web3-utils';
import validator from 'validator';
import etherutils from 'ethereumjs-util';

class HttpOfferCreate {
  constructor(url) {
    this.url = url;
  }

  getUrlAccount(account) {
    return '?account=' + account;
  }

  createOffer(address, expert, duration, websocket) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      if (web3Utils.isAddress(address)) {
        resolve();
      } else {
        reject(`${address} is not an ethereum address`);
      }
    })
      .then(
        () =>
          new Promise((resolve, reject) => {
            if (web3Utils.isAddress(expert)) {
              resolve();
            } else {
              reject(`${expert} is not an ethereum address`);
            }
          })
      )
      .then(
        () =>
          new Promise((resolve, reject) => {
            if (duration && duration > 10 && websocket) {
              const body = JSON.stringify({
                expert: expert,
                ambassador: address,
                settlementPeriodLength: Number(duration),
                websocketUri: websocket
              });
              resolve(body);
            } else {
              reject();
            }
          })
      )
      .then(body =>
        fetch(url + '/offers' + this.getUrlAccount(address), {
          headers: {
            'Content-Type': 'application/json'
          },
          method: 'post',
          body: body
        })
      )
      .then(response => {
        if (response.ok) {
          return response;
        }
        return new Promise(resolve => {
          resolve(response.json());
        }).then(json => {
          throw Error(json.message);
        });
      })
      .then(response => response.json())
      .then(body => body.result);
  }

  openOffer(key, offer, balance) {
    const url = this.url;
    new Promise((resolve, reject) => {
      if (validator.isUUID(offer.guid, 4)) {
        resolve();
      } else {
        reject('Invalid GUID');
      }
    })
      .then(
        () =>
          new Promise((resolve, reject) => {
            if (web3Utils.isAddress(offer.ambassador)) {
              resolve();
            } else {
              reject();
            }
          })
      )
      .then(
        () =>
          new Promise((resolve, reject) => {
            if (balance && balance > 0) {
              const offerState = [];
              offerState.push(0); // is close
              offerState.push(0); // sequence
              offerState.push(offer.ambassador); // ambassador address
              offerState.push(offer.expert); // expert address
              offerState.push(offer.msig); //  msig address
              offerState.push(balance); // new balance in nectar ambassador
              offerState.push(0); // balance in nectar expert
              offerState.push(0); // token address
              offerState.push(offer.guid); // A globally-unique identifer for the Listing.
              offerState.push(balance); // The Offer Amount.
              offerState.push(0); // Cryptographic hash of the Artifact.
              offerState.push(0); // The URI of the Artifact.
              offerState.push(0); // Engagement Deadline
              offerState.push(0); // Assertion Deadline
              offerState.push(0); // has the expert made commitment
              offerState.push(0); // “malicious” or “benign”
              offerState.push(0); // Information derived during Assertion generation

              const depositState = this.marshallState(offerState);
              const buff_key = etherutils.toBuffer(key);
              const state = web3Utils.toHex(depositState);
              let msg =
                '0x' +
                etherutils.keccak(etherutils.toBuffer(state)).toString('hex');
              msg =
                '0x' +
                etherutils
                  .hashPersonalMessage(etherutils.toBuffer(msg))
                  .toString('hex');
              const sig = etherutils.ecsign(etherutils.toBuffer(msg), buff_key);
              let r = '0x' + sig.r.toString('hex');
              let s = '0x' + sig.s.toString('hex');
              let v = sig.v;

              const body = JSON.stringify({
                state,
                v,
                r,
                s
              });
              resolve(body);
            } else {
              reject();
            }
          })
      )
      .then(body =>
        fetch(
          url +
            '/offers/' +
            offer.guid +
            '/open' +
            this.getUrlAccount(offer.ambassador),
          {
            headers: {
              'Content-Type': 'application/json'
            },
            method: 'post',
            body: body
          }
        )
      )
      .then(response => {
        if (response.ok) {
          return response;
        }
        return new Promise(resolve => {
          resolve(response.json());
        }).then(json => {
          throw Error(json.message);
        });
      })
      .then(response => response.json())
      .then(body => body.result);
  }

  marshallState(inputs) {
    var m = this.getBytes(inputs[0]);

    for (var i = 1; i < inputs.length; i++) {
      m += this.getBytes(inputs[i]).substr(2, this.getBytes(inputs[i]).length);
    }
    return m;
  }

  getBytes(input) {
    if (Buffer.isBuffer(input)) input = '0x' + input.toString('hex');
    if (66 - input.length <= 0) return web3Utils.toHex(input);
    return this.padBytes32(web3Utils.toHex(input));
  }

  padBytes32(data) {
    // TODO: check input is hex / move to TS
    let l = 66 - data.length;

    let x = data.substr(2, data.length);

    for (var i = 0; i < l; i++) {
      x = 0 + x;
    }
    return '0x' + x;
  }
}
export default HttpOfferCreate;
