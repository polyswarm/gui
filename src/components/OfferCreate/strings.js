const strings = {
  createBounty: 'Create Bounty',
  error: 'Something went wrong.',
  locked: 'Account was locked.',
  requestCreateBounty: 'Creating Bounty',
  reward: 'Maximum Reward (NCT)',
  duration: 'Duration (Blocks)',
  expert: 'Expert\'s Ethereum Address',
  openOffer: 'Open Offer Channel',
};
export default strings;
