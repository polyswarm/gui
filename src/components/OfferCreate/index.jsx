// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Uuid from 'uuid/v4';
import BigNumber from 'bignumber.js';
import web3Utils from 'web3-utils';
// Offer imports
import AnimatedInput from '../AnimatedInput';
import Button from '../Button';
import Header from '../Header';
// Component imports
import strings from './strings';
import HttpOfferCreate from './http';

class OfferCreate extends Component {
  constructor(props) {
    super(props);
    this.state = {
      reward_error: null,
      reward: null,
      duration: null,
      duration_error: null,
      expert: null,
      expert_error: null,
      nectar_error: null,
      nectar: null,
      websocket: null,
      websocket_error: null,
      creating: false
    };

    this.onClickHandler = this.onClickHandler.bind(this);
    this.onDurationChanged = this.onDurationChanged.bind(this);
    this.onExpertChanged = this.onExpertChanged.bind(this);
    this.onRewardChanged = this.onRewardChanged.bind(this);
    this.onWebsocketChanged = this.onWebsocketChanged.bind(this);
    this.addCreateOfferRequest = this.addCreateOfferRequest.bind(this);
    this.removeCreateOfferRequest = this.removeCreateOfferRequest.bind(this);
    this.validateFields = this.validateFields.bind(this);
  }

  componentDidMount() {
    const { props: { url } } = this;
    this.http = new HttpOfferCreate(url);
  }

  render() {
    const { state: { reward, reward_error, duration, duration_error, expert, expert_error, websocket, websocket_error, creating } } = this;
    const { props: {  wallet, address, requestsInProgress, onBackPressed } } = this;

    return (
      <div className='OfferCreate'>
        <Header title={strings.createOffer}
          requests={requestsInProgress}
          back={true}
          onBack={onBackPressed}
          address={address}
          wallet={wallet}/>
        <div className='OfferCreate-Content'>
          <div className='Offer-Values'>
            <h2>{strings.title}</h2>
            <AnimatedInput type='text'
              onChange={this.onExpertChanged}
              error={expert_error}
              placeholder={strings.expert}
              input_id='expert'/>
            <AnimatedInput type='number'
              onChange={this.onRewardChanged}
              error={reward_error}
              placeholder={strings.reward}
              input_id='reward'/>
            <AnimatedInput type='number'
              onChange={this.onDurationChanged}
              error={duration_error}
              placeholder={strings.duration}
              input_id='duration'/>
            <AnimatedInput type='text'
              onChange={this.onWebsocketChanged}
              error={websocket_error}
              placeholder={strings.websocket}
              input_id='websocket'/>
          </div>
          <div className='Offer-Create-Upload'>
            <Button
              disabled={!creating && (!reward || !duration || reward_error || duration_error || !expert || expert_error || !websocket || websocket_error) }
              onClick={this.onClickHandler}>
              {strings.openOffer}
            </Button>
          </div>
        </div>
      </div>
    );
  }

  onOfferPosted() {
    const {props: {onOfferPosted}} = this;
    if (onOfferPosted) {
      onOfferPosted();
    }
  }
  
  onClickHandler() {
    this.createOffer();
  }
  
  onDurationChanged(duration) {
    this.setState({duration: duration}, () => {
      this.validateFields();
    });
  }
  
  onExpertChanged(expert) {
    this.setState({expert: expert}, () => {
      this.validateFields();
    });
  }

  onRewardChanged(reward) {
    this.setState({reward: reward}, () => {
      this.validateFields();
    });
  }
  
  onWebsocketChanged(websocket) {
    this.setState({websocket: websocket}, () => {
      this.validateFields();
    });
  }

  createOffer() {
    const { state: {expert, expert_error, reward, reward_error, duration,
      duration_error, websocket, websocket_error} } = this;
    const { props: { addOffer, address, onOfferCreated, encryptionKey, token } } = this;

    const rewardWei = web3Utils.toWei(reward);

    const http = this.http;
    if (expert && reward && duration && websocket && !duration_error && !reward_error && !expert_error && !websocket_error) {
      this.setState({creating: true});
      const uuid = Uuid();
      this.addCreateOfferRequest(uuid);
      return new Promise(resolve => {
        if (onOfferCreated) {
          onOfferCreated();
        }
        resolve();
      })
        .then(() => http.createOffer(address, expert, Number(duration), websocket))
        .then(result => {
          if (addOffer) {
            addOffer(result, websocket, reward);
          }
          return result;
        })
        .then((result) => http.openOffer(encryptionKey, token, result, rewardWei))
        .catch(error => {
          this.setState({creating: false});
          let errorMessage;
          if (!error || !error.message || error.message.length === 0) {
            errorMessage = strings.error;
          } else {
            errorMessage = error.message;
          }
          this.setState({ error: errorMessage });

          //Update app
          const { props: { onError } } = this;
          if (onError) {
            onError(errorMessage);
          }
        })
        .then(() => {
          this.removeCreateOfferRequest(uuid);
        });
    } else {
      return null;
    }
  }

  addCreateOfferRequest(id) {
    const { addRequest } = this.props;
    if (addRequest) {
      addRequest(strings.requestCreateOffer, id);
    }
  }

  removeCreateOfferRequest(id) {
    const { removeRequest } = this.props;
    if (removeRequest) {
      removeRequest(strings.requestCreateOffer, id);
    }
  }

  validateFields() {
    const {state: {duration, reward,  expert, websocket}} = this;
    if (duration && duration < 10) {
      this.setState({duration_error: 'Duration below 10.'});
    } else if (duration && !Number.isInteger(Number(duration))) {
      this.setState({duration_error: 'Duration must be integer.'});
    } else {
      this.setState({duration_error: null});
    }

    const min = new BigNumber('0');
    if (reward && new BigNumber(reward).comparedTo(min) <= 0 ) {
      this.setState({reward_error: 'Reward below 0.'});
    } else {
      this.setState({reward_error: null});
    }

    if (websocket && !websocket.startsWith('ws://')) {
      this.setState({websocket_error: 'Value must be a websocket uri'});
    } else {
      this.setState({websocket_error: null});
    }

    if (expert && !web3Utils.isAddress(expert)) {
      this.setState({expert_error: 'Expert address must be a valid Ethereum address.'});
    } else {
      this.setState({expert_error: null});
    }
  }
}

OfferCreate.propTypes = {
  wallet: PropTypes.object,
  address: PropTypes.string,
  onError: PropTypes.func,
  onOfferCreated: PropTypes.func,
  addOffer: PropTypes.func,
  addRequest: PropTypes.func,
  removeRequest: PropTypes.func,
  url: PropTypes.string,
  onRequestWalletChange: PropTypes.func,
  encryptionKey: PropTypes.object,
};
export default OfferCreate;
