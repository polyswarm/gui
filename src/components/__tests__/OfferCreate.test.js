import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import OfferCreate from '../OfferCreate';
import HttpOfferCreate from '../OfferCreate/http';

const walletList = [
  {address:'author', nct: '1', eth: '1'},
  {address:'demo', nct: '1', eth: '1'},
  {address:'omed', nct: '1', eth: '1'}
];

const mockCreateOffer = jest.fn().mockImplementation(() => {
  return new Promise(resolve => resolve());
});

jest.mock('../OfferCreate/http', () => {
  // Works and lets you check for constructor calls:
  return jest.fn().mockImplementation(() => {
    return {
      createOffer: mockCreateOffer,
    };
  });
});

beforeEach(() => {
  localStorage.clear();
  jest.clearAllMocks();
  jest.setMock('react-transition-group', require('../__mocks__/react-transition-group'));
  HttpOfferCreate.mockClear();
  HttpOfferCreate.mockImplementation(() => {
    return {
      createOffer: mockCreateOffer,
    };
  });
});

it('renders without crashing', () => {
  const wrapper = render(<OfferCreate walletList={walletList}/>);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('disables the button when max reward is less than 0.0625', (done) =>{
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();
  
  instance.onDurationChanged('300');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('.006');

  wrapper.setState({}, () => {
    try {
      expect(wrapper.find('.Offer-Create-Upload').find('button').props().disabled).toBeTruthy();
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('disables the button when duration is 0', (done) => {
  const wrapper = mount(<OfferCreate  walletList={walletList}/>);
  const instance = wrapper.instance();
  
  instance.onDurationChanged('0');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('1');

  wrapper.setState({}, () => {
    try {
      expect(wrapper.find('.Offer-Create-Upload').find('button').props().disabled).toBeTruthy();
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('disables the button when duration is negative', (done) => {
  const wrapper = mount(<OfferCreate  walletList={walletList}/>);
  const instance = wrapper.instance();

  instance.onDurationChanged('-300');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('1');

  wrapper.setState({}, () => {
    try {
      expect(wrapper.find('.Offer-Create-Upload').find('button').props().disabled).toBeTruthy();
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('disables the button when expert address is not an ethereum address', (done) => {
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();

  instance.onDurationChanged('300');
  instance.onExpertChanged('0x000sdf7h');
  instance.onRewardChanged('1');

  wrapper.setState({}, () => {
    try {
      expect(wrapper.find('.Offer-Create-Upload').find('button').props().disabled).toBeTruthy();
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('enables the button when reward, duration and expert are filled with valid values', (done) => {
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();

  instance.onDurationChanged('300');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('1');

  wrapper.setState({}, () => {
    try {
      expect(wrapper.find('.Offer-Create-Upload').find('button').props().disabled).toBeFalsy();
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('opens the modal window when the button is clicked', (done) => {
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();

  instance.onDurationChanged('300');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('1');

  wrapper.setState({}, () => {
    wrapper.find('.Offer-Create-Upload').find('button').simulate('click');
    try {
      expect(wrapper.find('.ModalPassword')).toHaveLength(1);
      done();
    } catch (error) {
      done.fail(error);
    }
  });
});

it('calls createOffer when onWalletChangedHandler is called with true', () => {
  const createOffer = jest.spyOn(OfferCreate.prototype, 'createOffer');
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();

  instance.onDurationChanged('300');
  instance.onExpertChanged('0xAF8302a3786A35abEDdF19758067adc9a23597e5');
  instance.onRewardChanged('1');

  instance.onWalletChangeHandler(true);

  expect(createOffer).toHaveBeenCalledTimes(1);
});

it('calls http.createOffer when onWalletChangedHandler is called with true', (done) => {
  const wrapper = mount(<OfferCreate walletList={walletList}/>);
  const instance = wrapper.instance();

  wrapper.setState({
    duration: '300',
    reward: '1',
    expert: '0xAF8302a3786A35abEDdF19758067adc9a23597e5'
  }, () => {
    instance.createOffer()
      .then(() => {
        try {
          expect(mockCreateOffer).toHaveBeenCalledWith(
            '0xAF8302a3786A35abEDdF19758067adc9a23597e5',
            '1000000000000000000',
            300);
          done();
        } catch (error) {
          done.fail(error);
        }
      });
  });
});

it('calls prop addOffer if posted successfully', (done) => {
  const addOffer = jest.fn();
  const wrapper = mount(<OfferCreate 
    addOffer={addOffer}
    walletList={walletList}/>);
  const instance = wrapper.instance();

  wrapper.setState({
    duration: '300',
    reward: '1',
    expert: '0xAF8302a3786A35abEDdF19758067adc9a23597e5'
  }, () => {
    instance.createOffer()
      .then(() => {
        try {
          expect(addOffer).toHaveBeenCalledTimes(1);
          done();
        } catch (error) {
          done.fail(error);
        }
      });
  });
});

it('calls prop onError when http.createOffer fails', (done) => {
  const mockBadCreateOffer = jest.fn().mockImplementation(() => {
    return new Promise((result, reject) => {
      const error = {
        status: 401,
      };
      reject(error);
    });
  });
  HttpOfferCreate.mockImplementation(() => {
    return {
      createOffer: mockBadCreateOffer
    };
  });
  const onError = jest.fn();
  const wrapper = mount(<OfferCreate 
    onError={onError}
    walletList={walletList}/>);
  const instance = wrapper.instance();

  wrapper.setState({
    duration: '300',
    reward: '1',
    expert: '0xAF8302a3786A35abEDdF19758067adc9a23597e5'
  }, () => {
    instance.createOffer().then(() => {
      try {
        expect(onError).toHaveBeenCalledTimes(1);
        done();
      } catch (error) {
        done.fail(error);
      }
    });
  });
});