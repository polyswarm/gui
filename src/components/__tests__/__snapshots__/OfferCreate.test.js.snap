// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="OfferCreate"
>
  <header
    class="Header"
  >
    <div
      class="Header-Row"
    >
      <div
        class="Header-Title"
      >
        <img
          alt="PolySwarm logo"
          class="Header-Logo"
          src="../public/img/back-arrow.svg"
        >
          
        </img>
        <h3>
          Create Offer
        </h3>
        <span
          class="RequestSpinner"
        >
          
        </span>
      </div>
      <div
        class="Header-Actions"
      >
        
      </div>
    </div>
    <div
      class="Header-Address"
    >
      
    </div>
  </header>
  <div
    class="OfferCreate-Content"
  >
    <div
      class="Offer-Values"
    >
      <h2>
        Enter Expert, max reward and duration.
      </h2>
      <div
        class="AnimatedInput"
      >
        <div>
          
        </div>
        <input
          class="AnimatedInput-Input"
          id="expert"
          type="text"
          value=""
        >
          
        </input>
      </div>
      <div
        class="AnimatedInput"
      >
        <div>
          
        </div>
        <input
          class="AnimatedInput-Input"
          id="reward"
          type="number"
          value=""
        >
          
        </input>
      </div>
      <div
        class="AnimatedInput"
      >
        <div>
          
        </div>
        <input
          class="AnimatedInput-Input"
          id="duration"
          type="number"
          value=""
        >
          
        </input>
      </div>
      <div
        class="AnimatedInput"
      >
        <div>
          
        </div>
        <input
          class="AnimatedInput-Input"
          id="port"
          type="number"
          value=""
        >
          
        </input>
      </div>
    </div>
    <div
      class="Offer-Create-Upload"
    >
      <button
        class="Button"
        disabled=""
      >
        Open Offer Channel
      </button>
    </div>
  </div>
</div>
`;
